woocommerce/woocommerce
# Add blueprint behind a feature flag

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





This PR adds pdibGW-3lQ-p2 with a feature flag.

Notable changes:

1. Added Blueprint as a package in [packages/php/blueprint](https://github.com/woocommerce/woocommerce/tree/add/blueprint-feature-flag/packages/php/blueprint) to decouple it from WooCommerce so that other plugins can extend/customize it.
2. WooCommerce extends the package and codes live in [client/blueprint](https://github.com/woocommerce/woocommerce/tree/add/blueprint-feature-flag/plugins/woocommerce-admin/client/blueprint) for the frontend, [src/Admin/Features/Blueprint](https://github.com/woocommerce/woocommerce/tree/add/blueprint-feature-flag/plugins/woocommerce/src/Admin/Features/Blueprint) for the backend.






### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Create a new JN site with this branch
4. Confirm the core profiler does not have `Builder setup` button yet. Do not proceed or skip the profiler yet.
5. Open a new window and navigate to the JN site admin.
6. Navigate to `Tools -> WCA Test Helper -> Features` and enable `blueprint` feature flag.
7. Go back to the core profiler page and refresh.
8. Confirm you have `Builder setup` button. Skip or complete the core profiler.
9. Navigate to `WooCommerce -> Settings -> Advanced` and confirm `blueprint` subsection has been added.
10. Click on the `blueprint` link.
11. Confirm the settings page renders without an error.


### Testing Export

1. Use the same JN site.
2. Navigate to `WooCommerce -> Settings -> Advanced -> Blueprint`
3. Click on `Export`
4. Save the file on your computer.
5. Create a new JN site with `blueprint` feature enabled.
6. Click on `builder setup` button on the core profiler page.
7. Choose the exported JSON and click on `Import`
8. Confirm you're redirected to WooCommerce Home shortly.


### Testing Zip Export

1. Go to woo.com and sign-in.
2. Navigate to https://woocommerce.com/my-account/downloads/
3. Download a plugin.
4. Use the same JN site.
5. Install and activate the plugin.
6. Navigate to `WooCommerce -> Settings -> Advanced -> Blueprint`
7. Click on `Export as a zip (Experimental)` checkbox
8. Click on `Export` button.
9. It should prompt a zip file for this time.
10. Save the zip file locally.
11. Create a new JN site with `blueprint` feature enabled.
12. Click on `builder setup` button on the core profiler page.
13. Choose the exported zip file and click on `Import`
14. Confirm you're redirected to WooCommerce Home shortly.
15. Navigate to `Plugins` page and verify that the plugin from woocommerce.com is present.

### Testing CLI Commands

1. Use the same JN site.
2. SSH into the site.
3. Test the following commands

Import: ```wp wc blueprint import path-to-woo-blueprint.json```

Export: ```wp wc blueprint export save-to-path.json```

Optionally, you can modify values in the exported file and confirm they are imported successfully





### Changelog entry







<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 
Add blueprint export/import behind a feature flag for testing purposes.

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
